### PR TITLE
fix(ci): Generate per-feed package lists in server build pipeline

### PR DESCRIPTION
**Pull request by alexvy86**
## Description

The server-related pipelines (`server - routerlicious`, `server - historian`, `server - gitrest`) that generate npm packages currently don't generate files with lists of packages to publish to each feed, like the client release group pipeline does. This is making `server - routerlicious` incorrectly try to publish all packages to the build and dev feeds, and sometimes fail because the second one upstreams the first one and sometimes (depending on timing) it complains about publishing a package to it when one of its upstreams already has that version.

Copied the necessary logic from `scripts/pack-packages.sh` because the fact that the server pipelines do their build inside a Docker container complicates using the script. Would be nice to consolidate later.

Also moved from one ADO system variable to another which more clearly reflects what we're putting in it.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

Confirmed that the pipeline run for this PR [does include the packages-per-feed files in its published artifacts](https://dev.azure.com/fluidframework/public/_build/results?buildId=213543&view=artifacts&pathAsName=false&type=publishedArtifacts) (under `pack`). Compare to [one without this change](https://dev.azure.com/fluidframework/public/_build/results?buildId=213448&view=artifacts&pathAsName=false&type=publishedArtifacts).

Also confirmed that the other server pipelines (historian, gitrest) publish the package lists correctly.

### To ponder

We recently added the [`isReleaseGroup` parameter](https://github.com/alexvy86/FluidFramework/blob/8728c12da1f7a4a662b12102b287a2e6649e727b/tools/pipelines/server-routerlicious.yml#L114) set to `true` in the build client and server routerlicious pipelines, with the only intention (so far) of adding a stage of publishing packages to the dev feed (in addition to the build feed), but I realized that only the client release group (for now, at least) publishes separate sets of packages to each feed. So we could also change that parameter to `false` for the server - routerlicious pipeline, and the issue with it publishing the same packages to the dev and build feeds would go away without having to generate package lists for it. That feels weird given that it _is_ a release group... but so are [historian](https://github.com/microsoft/FluidFramework/tree/main/server/historian) and [gitrest](https://github.com/microsoft/FluidFramework/tree/main/server/gitrest) (albeit small ones) and we did not set the `isReleaseGroup` parameter to `true` for them. 

In short, there's still some thinking and cleanup to do around our logic for handling this, but the package lists for both of those (and server/routerlicious) look correct and thus the packages that would end up published seem correct too.

### CODE REVIEW

File: tools/pipelines/templates/build-docker-service.yml [hunk 1]
```diff
@@ -293,16 +293,41 @@ stages:
...
+              flub list --no-private --releaseGroup $RELEASE_GROUP --tarball --feed public > $STAGING_PATH/pack/packagePublishOrder-public.txt
+              flub list --no-private --releaseGroup $RELEASE_GROUP --tarball --feed internal-build > $STAGING_PATH/pack/packagePublishOrder-internal-build.txt
+              flub list --no-private --releaseGroup $RELEASE_GROUP --tarball --feed internal-dev > $STAGING_PATH/pack/packagePublishOrder-internal-dev.txt
+              flub list --no-private --releaseGroup $RELEASE_GROUP --tarball --feed internal-test > $STAGING_PATH/pack/packagePublishOrder-internal-test.txt
 
         - task: PublishPipelineArtifact@1
           displayName: Publish Artifact - pack
           inputs:
-            targetPath: '$(System.DefaultWorkingDirectory)/pack'
+            targetPath: '$(Build.ArtifactStagingDirectory)/pack'
```
Comment (alexvy86) [inline tools/pipelines/templates/build-docker-service.yml:330]: Updated everywhere where it relates to creating artifacts because I think it makes intentions clearer; that's where someone familiar with ADO pipelines would expect to see anything that is going to be published as artifacts later on.

**Comment by alexvy86** [comment]
@jikim-msft @Josmithr this is what I mentioned in standup.

File: tools/pipelines/templates/build-docker-service.yml [hunk 2]
```diff
@@ -293,16 +293,41 @@ stages:
...
+
         - task: Docker@0
           displayName: Pack
           inputs:
             action: 'Run a Docker command'
-            customCommand: 'run -v $(System.DefaultWorkingDirectory)/pack/scoped:/usr/src/pack -t $(baseContainerTag) pnpm lerna exec --no-private -- mv `npm pack` /usr/src/pack'
+            customCommand: 'run -v $(Build.ArtifactStagingDirectory)/pack/scoped:/usr/src/pack -t $(baseContainerTag) pnpm lerna exec --no-private -- mv `npm pack` /usr/src/pack'
+
+        # Need to generate package lists manually, because we don't have access to the scripts/pack-packages.sh
+        # script that build-npm-package.yml uses.
```
Comment (tylerbutler) [inline tools/pipelines/templates/build-docker-service.yml:312]: This isn't quite true. We now have a way to reference any path in the repo within a Dockerfile. I'll dig up an example.

Comment (tylerbutler) [reply tools/pipelines/templates/build-docker-service.yml:312]: Here's the example: https://github.com/microsoft/FluidFramework/blob/58eebc1ca77041c66cbc54c82e0a400e47f9533c/server/routerlicious/Dockerfile#L25-L27

the "root" is the root of the repo. You can use this to copy anything in the repo into the image for use. I discovered the "build context" capability [somewhat recently](https://github.com/microsoft/FluidFramework/pull/17605).

Anyway, this should let you keep more of the build logic in the Dockerfile.

Comment (alexvy86) [reply tools/pipelines/templates/build-docker-service.yml]: Giving it a shot

File: tools/pipelines/templates/build-docker-service.yml [hunk 3]
```diff
@@ -93,7 +93,7 @@ parameters:
 # Any additional build arguments to pass to the docker build command.
 - name: additionalBuildArguments
   type: string
-  default: ""
+  default: "--build-context root=."
```
Comment (Josmithr) [inline tools/pipelines/templates/build-docker-service.yml:96]: Can we document this default?

Comment (alexvy86) [reply tools/pipelines/templates/build-docker-service.yml:96]: I'm actually very tempted to revert all the changes[8728c12](https://github.com/microsoft/FluidFramework/pull/18445/commits/8728c12da1f7a4a662b12102b287a2e6649e727b), I haven't been able to make them work yet.

**Comment by alexvy86** [comment]
I tried using `pack-packages.sh` inside the container but it proved too problematic so I'm sticking with the approach that worked for now. Some issues I ran across:

- ["flub: command not found"](https://dev.azure.com/fluidframework/public/_build/results?buildId=214348&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=a1aa9649-a94b-5ac4-3f5e-9bb6223edb04&l=37). Fixed by using `pnpm flub` instead of `flub` inside `scripts/pack-packages.sh`. This might be a good idea regardless of this PR.
- ["fatal: not a git repository"](https://dev.azure.com/fluidframework/public/_build/results?buildId=214392&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=a1aa9649-a94b-5ac4-3f5e-9bb6223edb04&l=43). Which makes sense, because when copying files to the container we're not bringing in `.git`.

That approach also required updating the separate Dockerfiles for the server release groups (routerlicious, historian, gitrest) so they all copied the sh script (which meant they all now used the `--build-context` argument) and adding a step to install `git` in the gitrest one.

At that point I decided it's becoming too big of a change to keep pushing in this direction, so I think issuing the working alternative makes more sense for now, and eventually we should just stop doing the builds for server packages inside docker (following chats I had with Curtis in the past). The docker images should install the packages from wherever a "normal build" publishes them.

File: tools/pipelines/templates/build-docker-service.yml (hunk 3 above)
Comment (alexvy86) [reply tools/pipelines/templates/build-docker-service.yml]: Reverted this.

**Comment by tylerbutler** [comment]
> At that point I decided it's becoming too big of a change to keep pushing in this direction

Good call! Building outside docker is probably much more achievable these days; best to focus our energy there.

**Review by Josmithr — APPROVED** [review]

File: tools/pipelines/templates/build-docker-service.yml [hunk 4]
```diff
@@ -293,16 +293,42 @@ stages:
...
-            customCommand: 'run -v $(System.DefaultWorkingDirectory)/pack/scoped:/usr/src/pack -t $(baseContainerTag) pnpm lerna exec --no-private -- mv `npm pack` /usr/src/pack'
+            customCommand: 'run -v $(Build.ArtifactStagingDirectory)/pack/scoped:/usr/src/pack -t $(baseContainerTag) pnpm lerna exec --no-private -- mv `npm pack` /usr/src/pack'
+
+        # Generating package lists manually because using scripts/pack-packages.sh (like build-npm-package.yml does)
+        # inside the Docker container is proving problematic.
+        # See https://github.com/microsoft/FluidFramework/pull/18445 for details.
+        - task: Bash@3
+          displayName: Generate package lists
+          env:
+            RELEASE_GROUP: ${{ parameters.tagName }}
```
[the remark this answers is not in the document]
Comment (alexvy86) [reply tools/pipelines/templates/build-docker-service.yml:317]: I think it actually started as the value we use to generate tags in the git repo, and not only for release groups, so it makes a lot of sense there.

image

We then added more uses for the same values in the case of release groups, like `flub` [using them](https://github.com/alexvy86/FluidFramework/blob/1210d1200b152e18c3829e6d38a82fc82d6b1434/build-tools/packages/build-cli/src/releaseGroups.ts#L23-L29) to know which scope it wants to do something on.
